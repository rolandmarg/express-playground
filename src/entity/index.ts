import { User } from './User';
import { Meeting } from './Meeting';

export { User, Meeting };

export default [User, Meeting];
