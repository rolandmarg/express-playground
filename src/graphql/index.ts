export * from './apollo';

export * as queries from './query';
export * as mutations from './mutation';
export * as fragments from './fragments';
