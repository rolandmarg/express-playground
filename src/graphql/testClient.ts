import {
  createTestClient,
  ApolloServerTestClient,
} from 'apollo-server-testing';
import { apollo } from './apollo';

let client: ApolloServerTestClient;

const initClient = async () => {
  const server = await apollo();

  client = createTestClient(server);

  return client;
};

type Query = Parameters<typeof client.query>[0];
type Mutation = Parameters<typeof client.mutate>[0];

export const query = async (query: Query) => {
  if (!client) {
    await initClient();
  }

  return client.query(query);
};

export const mutate = async (mutation: Mutation) => {
  if (!client) {
    await initClient();
  }

  return client.mutate(mutation);
};
