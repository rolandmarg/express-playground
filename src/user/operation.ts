import { gql } from 'apollo-server-express';

export const userFragment = gql`
  fragment UserFields on User {
    id
    email
    createdAt
  }
`;

export const meQuery = gql`
  query Me {
    me {
      ...UserFields
    }
  }
  ${userFragment}
`;

export const userQuery = gql`
  query User($id: ID!) {
    user(id: $id) {
      ...UserFields
    }
  }
  ${userFragment}
`;

export const usersQuery = gql`
  query Users {
    users {
      ...UserFields
    }
  }
  ${userFragment}
`;
