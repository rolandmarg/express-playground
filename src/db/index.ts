import { createConnection, Connection } from 'typeorm';
import env from '../env';

import { User } from './entity/User';
import { Meeting } from './entity/Meeting';
import { Provider } from './entity/Provider';

export * from 'typeorm';
export * as meetingRepo from './repository/meeting';

export { User, Meeting, Provider };

let connection: Connection;
export const connect = async () => {
  connection = await createConnection({
    type: 'postgres',
    url: env.DB_URL,
    synchronize: env.NODE_ENV !== 'production',
    entities: [User, Meeting, Provider],
  });

  return connection;
};

export const close = async () => {
  return connection.close();
};
