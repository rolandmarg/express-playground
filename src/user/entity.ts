import {
  Entity,
  PrimaryGeneratedColumn,
  Column,
  CreateDateColumn,
  OneToMany,
} from 'typeorm';
import { Provider } from '../provider/entity';
import { ObjectType, Field, ID } from 'type-graphql';

@ObjectType()
@Entity({ name: 'users' })
export class User {
  @Field(() => ID)
  @PrimaryGeneratedColumn()
  id!: number;

  @Field()
  @Column({ unique: true })
  email!: string;

  @Field()
  @CreateDateColumn()
  createdAt!: Date;

  @Field(() => [Provider])
  @OneToMany(() => Provider, (provider) => provider.user)
  providers!: Provider[];
}
