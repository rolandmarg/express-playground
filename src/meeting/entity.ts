import { Entity, PrimaryGeneratedColumn, Column, Index } from 'typeorm';
import { ObjectType, Field, ID } from 'type-graphql';
import { Connection } from '../graphql/types';

@ObjectType()
@Entity({ name: 'meetings' })
export class Meeting {
  @Field(() => ID)
  @PrimaryGeneratedColumn()
  id!: number;

  @Field()
  @Column()
  title!: string;

  @Field()
  @Index()
  @Column()
  startsAt!: Date;

  @Field()
  @Column()
  endsAt!: Date;
}

@ObjectType()
export class MeetingConnection extends Connection(Meeting) {}
