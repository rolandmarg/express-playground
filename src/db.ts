import { createConnection, Connection } from 'typeorm';
import {
  NODE_ENV,
  DB_DATABASE,
  DB_HOST,
  DB_PASSWORD,
  DB_PORT,
  DB_USER,
} from './utils';
import { User } from './user/entity';
import { Meeting } from './meeting/entity';
import { Provider } from './provider/entity';

let connection: Connection;

export const connect = async () => {
  connection = await createConnection({
    type: 'postgres',
    host: DB_HOST,
    port: DB_PORT,
    username: DB_USER,
    password: DB_PASSWORD,
    database: DB_DATABASE,
    synchronize: NODE_ENV !== 'production',
    entities: [User, Meeting, Provider],
  });

  return connection;
};

export const close = async () => {
  return connection.close();
};
