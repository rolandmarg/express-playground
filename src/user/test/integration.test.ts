import { connect, close } from '../../db';
import { query } from '../../graphql/testClient';
import { userQuery, usersQuery } from '../operation';
import { getRepository, Repository } from 'typeorm';
import { User } from '../entity';

describe('User operations', () => {
  let userRepo: Repository<User>;

  beforeAll(async () => {
    await connect();

    userRepo = getRepository(User);
  });

  afterAll(async () => {
    await close();
  });

  beforeEach(async () => {
    await userRepo.delete({});
  });

  it('should fetch all users', async () => {
    const user = userRepo.create();
    user.email = 'test@gmail.com';

    await userRepo.save(user);

    const res = await query({ query: usersQuery });

    expect(res).toMatchSnapshot({
      data: {
        users: [
          {
            id: expect.any(String),
            createdAt: expect.any(String),
            email: user.email,
          },
        ],
      },
    });
  });

  it('should fetch empty list of users', async () => {
    const res = await query({ query: usersQuery });

    expect(res).toMatchSnapshot();
  });

  it('should fetch specific user', async () => {
    const user = userRepo.create();
    user.email = 'test@gmail.com';

    await userRepo.save(user);

    const res = await query({
      query: userQuery,
      variables: { id: user.id },
    });

    expect(res).toMatchSnapshot({
      data: {
        user: {
          id: expect.any(String),
          createdAt: expect.any(String),
          email: user.email,
        },
      },
    });

    expect(+res.data?.user.id).toBe(user.id);
  });
});
