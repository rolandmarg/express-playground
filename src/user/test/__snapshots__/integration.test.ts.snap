// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`User operations should fetch all users 1`] = `
Object {
  "data": Object {
    "users": Array [
      Object {
        "createdAt": Any<String>,
        "email": "test@gmail.com",
        "id": Any<String>,
      },
    ],
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`User operations should fetch empty list of users 1`] = `
Object {
  "data": Object {
    "users": Array [],
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`User operations should fetch specific user 1`] = `
Object {
  "data": Object {
    "user": Object {
      "createdAt": Any<String>,
      "email": "test@gmail.com",
      "id": Any<String>,
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;
